vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
 }
 
 facebook::jsi::Array ChakraRuntime::getPropertyNames(const facebook::jsi::Object &object) {
-  JsValueRef propertyNamesArrayRef;
-  VerifyJsErrorElseThrow(JsGetOwnPropertyNames(GetChakraObjectRef(object), &propertyNamesArrayRef));
+  // Handle to the null JS value.
+  ChakraObjectRef jsNull = ToChakraObjectRef(facebook::jsi::Value::null());
+
+  // Handle to the Object constructor.
+  ChakraObjectRef objectConstructor = GetProperty(GetChakraObjectRef(global()), "Object");
+
+  // Handle to the Object.prototype Object.
+  ChakraObjectRef objectPrototype = GetProperty(objectConstructor, "prototype");
+
+  // Handle to the Object.prototype.propertyIsEnumerable() Function.
+  ChakraObjectRef objectPrototypePropertyIsEnumerable = GetProperty(objectPrototype, "propertyIsEnumerable");
+
+  std::vector<ChakraObjectRef> enumerablePropNames{};
+  ChakraObjectRef currentObject = GetChakraObjectRef(object);
 
-  JsPropertyIdRef propertyId;
-  VerifyJsErrorElseThrow(JsGetPropertyIdFromName(L"length", &propertyId));
-  JsValueRef countRef;
-  VerifyJsErrorElseThrow(JsGetProperty(propertyNamesArrayRef, propertyId, &countRef));
-  int count;
+  // We have a small optimization here where we stop traversing the prototype
+  // chain as soon as we hit Object.prototype. However, we still need to check
+  // for null here, as one can create an Object with no prototype through
+  // Object.create(null).
+  while (!CompareJsValues(currentObject, objectPrototype) && !CompareJsValues(currentObject, jsNull)) {
+    JsValueRef propNamesRef = JS_INVALID_REFERENCE;
+    VerifyJsErrorElseThrow(JsGetOwnPropertyNames(currentObject, &propNamesRef));
+    ChakraObjectRef propNames(propNamesRef);
 
-  VerifyJsErrorElseThrow(JsNumberToInt(countRef, &count));
+    int numPropNames = ToInteger(GetProperty(propNames, "length"));
 
-  auto result = createArray(count);
-  for (int i = 0; i < count; i++) {
-    JsValueRef index;
-    VerifyJsErrorElseThrow(JsIntToNumber(i, &index));
-    JsValueRef propertyName;
-    VerifyJsErrorElseThrow(JsGetIndexedProperty(propertyNamesArrayRef, index, &propertyName));
+    for (int i = 0; i < numPropNames; ++i) {
```

Comment:
>for [](start = 4, length = 3)

(nit) can you please add a small comment describing what this loop does? We are getting propNames on "currentObject", but then we are switching "currentObject" inside of loop.